pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -430,6 +438,171 @@ func consumerRefMatches(consumer *corev1.ObjectReference, machine *machinev1.Mac
+	// CA cert for the URL's https connection)
+	userDataSecret := corev1.Secret{}
+	key = client.ObjectKey{
+		Name:      config.UserData.Name,
+		Namespace: config.UserData.Namespace,
+	}
+	err = a.client.Get(ctx, key, &userDataSecret)
+	if err != nil {
+		log.Printf("Cannot get %s secret: %w", config.UserData.Name, err)
+		return fmt.Errorf("Cannot get %s secret: %w", config.UserData.Name, err)
+	}
+	encodedUserDataBytes := userDataSecret.Data["userData"]
+	var userData []byte
+	userData, err = base64.StdEncoding.DecodeString(string(encodedUserDataBytes))
+	if err != nil {
+		log.Printf("Error base64 decoding userData : %s", err.Error())
+		return fmt.Errorf("error base64 decoding userData from %s : %w", config.UserData.Name, err)
+	}
+
+	// Define an ignition data struct to unmarshal incoming json
+	type IgnitionData struct {
+		Ignition struct {
+			Config struct {
+				Append []struct {
+					Source string
+				}
+			}
+			Security struct {
+				TLS struct {
+					CertificateAuthorities []struct {
+						Source string
+					}
+				}
+			}
+		}
+	}
+	var ignData IgnitionData
+
+	if err := json.Unmarshal(userData, &ignData); err != nil {
+		log.Printf("Cannot unmarshal ignition config from userData : %s, %w", config.UserData.Name, err)
+		return fmt.Errorf("Cannot unmarshal ignition config from userData : %s, %w", config.UserData.Name, err)
+	}
+
+	var ignitionURL, caCertRaw string
+	// Fetch IgnitionURL and the CA cert
+	if len(ignData.Ignition.Config.Append) > 0 && len(ignData.Ignition.Config.Append[0].Source) > 0 {
+		ignitionURL = ignData.Ignition.Config.Append[0].Source
+	} else {
+		// We didn't get any URL
+		err := fmt.Errorf("error extracting full Ignition URL from content of %s", config.UserData.Name)
+		log.Printf("URL not found : %s", err.Error())
+		return err
+	}
+
+	if len(ignData.Ignition.Security.TLS.CertificateAuthorities) > 0 && len(ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source) > 0 {
+		caCertRaw = ignData.Ignition.Security.TLS.CertificateAuthorities[0].Source
+	} else {
+		// We didn't get the URL's Cert
+		err := fmt.Errorf("CA cert for full Ignition URL was not specified in %s", config.UserData.Name)
+		log.Printf("Cert not found in ignition data: %s", err.Error())
+		return err
+	}
+	caCertB64 := strings.TrimPrefix(caCertRaw, "data:text/plain;charset=utf-8;base64,")
+
+	caCert, err := base64.StdEncoding.DecodeString(caCertB64)
+	if err != nil {
+		log.Printf("Could not base64 decode caCert: %s", err.Error())
+		return fmt.Errorf("could not base64 decode caCert in %s: %w", config.UserData.Name, err)
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+	client := retryablehttp.NewClient()
+	client.HTTPClient.Transport = &http.Transport{
+		TLSClientConfig: &tls.Config{
+			RootCAs: caCertPool,
+		},
+	}
+
+	// Get the full ignition from the URL
+	resp, err := client.Get(ignitionURL)
+	if err != nil || resp.StatusCode != http.StatusOK {
```

Comment:
What do you think about handling the error and response codes separately? The docs for the standard library's client say "A non-2xx response doesn't cause an error." Not sure if the same applies to this library, but seems likely.

It would look like this: `if err != nil` log and return it. Then if the response code is unexpected, log and handle that accordingly. Seems simpler, and I'm not sure what value there is in evaluating the error and response code together. WDYT?